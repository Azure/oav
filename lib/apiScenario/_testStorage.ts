import { inversifyGetInstance } from "../inversifyUtils";
import "reflect-metadata";
import { PostmanCollectionGenerator } from "./postmanCollectionGenerator";

// oav generate-postmanCollection testScenario --env
const main = async () => {
  const env = {
    subscriptionId: "db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    location: "eastasia",
    scope: "subscriptions/db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    resourceId: "mysourceid",
    tenantId: "<your tenantId>",
    client_id: "<your client_id>",
    client_secret: "<your client secret>",
    resourceGroupName: "ruowan-test",
    deploymentName: "createVMSS",
  };

  const postmanCollectionGenerator = inversifyGetInstance(PostmanCollectionGenerator, {
    name: "storage",
    fileRoot: "/home/ruowan/work/azure-rest-api-specs/specification/storage/resource-manager",
    swaggerFilePaths: [
      "Microsoft.Storage/stable/2019-06-01/storage.json",
      "Microsoft.Storage/stable/2019-06-01/blob.json",
    ],
    testDef: "Microsoft.Storage/stable/2019-06-01/test-scenairos/createStorageAccount.yaml",
    env: env,
    outputFolder: "./fixtures",
    generateCollection: true,
    runCollection: false,
  });

  try {
    await postmanCollectionGenerator.GenerateCollection();
  } catch (e) {
    console.log(e.message, e.stack);
  }
};

console.time("TestLoad");
console.log("Start");

main().finally(() => {
  console.timeEnd("TestLoad");
});
