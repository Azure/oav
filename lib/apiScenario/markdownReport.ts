import { readFileSync } from "fs";
import path from "path";
import Handlebars from "handlebars";
import * as hd from "humanize-duration";
import moment from "moment";
import { ResponseDiffItem, RuntimeError, StepResult, TestScenarioResult } from "./reportGenerator";

const spaceReg = /(\n|\t|\r)/gi;

const commonHelper = (opts: HelperOpts) => ({
  renderPlain: (s: string) => s,
  renderWhitespace: (n: number) => "&nbsp;".repeat(n),
  renderUri: (s: string) => `${path.join(opts.swaggerRootDir, s)}`,
  renderSymbol: (result: ResultState) => `${resultStateSymbol[result]}`,
  renderScenarioTitle: (ts: TestScenarioMarkdownResult) => {
    let s = `${ts.testScenarioName}`;
    if (ts.failedStepsCount <= 0 && ts.fatalStepsCount <= 0) {
      return s;
    }
    s = `${s}: `;
    if (ts.fatalStepsCount > 0) {
      s = `${s}${ts.fatalStepsCount} Fatal Step(s)`;
    }

    if (ts.fatalStepsCount > 0 && ts.failedStepsCount > 0) {
      s = `${s}, `;
    }

    if (ts.failedStepsCount > 0) {
      s = `${s} ${ts.failedStepsCount} Failed Step(s)`;
    }

    return s;
  },
  renderStepTitle: (ts: TestScenarioMarkdownStepResult) => {
    let s = `${ts.stepName}`;
    if (ts.failedErrorsCount <= 0 && ts.fatalErrorsCount <= 0) {
      return s;
    }
    s = `${s}: `;
    if (ts.fatalErrorsCount > 0) {
      s = `${s}${ts.fatalErrorsCount} Fatal Error(s)`;
    }

    if (ts.fatalErrorsCount > 0 && ts.failedErrorsCount > 0) {
      s = `${s}, `;
    }

    if (ts.failedErrorsCount > 0) {
      s = `${s} ${ts.failedErrorsCount} Validation Error(s)`;
    }

    return s;
  },
  renderDuration: (start: Date, end: Date) =>
    `${hd.default(moment.duration(moment(end).diff(moment(start))).asMilliseconds())}`,
  shouldReportError: (sr: TestScenarioMarkdownStepResult) =>
    sr.failedErrorsCount + sr.fatalErrorsCount > 0,
  renderFatalErrorCode: (e: RuntimeError) => `[${e.code}]()`,
  renderFatalErrorDetail: (e: RuntimeError) => `${e.message.replace(spaceReg, " ")}`,
  renderDiffErrorCode: (e: ResponseDiffItem) => `[${e.code}]()`,
  renderDiffErrorDetail: (e: ResponseDiffItem) => `${e.message.replace(spaceReg, " ")}`,
});

type ResultState = keyof typeof ResultStateStrings;

const ResultStateStrings = {
  fatal: `Fatal`,
  failed: `Failed`,
  succeeded: `Succeeded`,
  warning: `Warning`,
};

export const resultStateSymbol: { [key in ResultState]: string } = {
  fatal: "❌",
  failed: "❌",
  succeeded: "️✔️",
  warning: "⚠️",
};

interface TestScenarioMarkdownStepResult {
  stepName: string;
  result: ResultState;
  exampleFilePath?: string;
  correlationId?: string;
  operationId: string;
  fatalErrorsCount: number;
  failedErrorsCount: number;
  warningErrorsCount: number;
  runtimeError?: RuntimeError[];
  responseDiffResult?: ResponseDiffItem[];
}

interface TestScenarioMarkdownResult {
  testScenarioName: string;
  result: ResultState;
  swaggerFilePaths: string[];
  startTime: Date;
  endTime: Date;
  runId: string;
  fatalStepsCount: number;
  failedStepsCount: number;
  warningStepsCount: number;
  steps: TestScenarioMarkdownStepResult[];
}

interface HelperOpts {
  swaggerRootDir: "root";
}

export const compileHandlebarsTemplate = <T>(fileName: string, opts: HelperOpts) => {
  const generationViewTemplate = readFileSync(
    path.join(__dirname, "templates", fileName)
  ).toString();
  const templateDelegate = Handlebars.compile<T>(generationViewTemplate, { noEscape: true });
  const helpers = commonHelper(opts);
  return (data: T) => templateDelegate(data, { helpers });
};

const generateMarkdownReportView = compileHandlebarsTemplate<TestScenarioMarkdownResult>(
  "markdownReport.handlebars",
  {
    swaggerRootDir: "root",
  }
);

const generateJUnitCaseReportView = compileHandlebarsTemplate<TestScenarioMarkdownStepResult>(
  "junitCaseReport.handlebars",
  {
    swaggerRootDir: "root",
  }
);

const stepIsFatal = (sr: StepResult) => sr.runtimeError && sr.runtimeError.length > 0;
const stepIsFailed = (sr: StepResult) => sr.responseDiffResult && sr.responseDiffResult.length > 0;

const asMarkdownStepResult = (sr: StepResult): TestScenarioMarkdownStepResult => {
  let result: ResultState = "succeeded";
  if (stepIsFatal(sr)) {
    result = "fatal";
  } else if (stepIsFailed(sr)) {
    result = "failed";
  }

  const r: TestScenarioMarkdownStepResult = {
    result,
    fatalErrorsCount: sr.runtimeError ? sr.runtimeError.length : 0,
    failedErrorsCount: sr.responseDiffResult ? sr.responseDiffResult.length : 0,
    warningErrorsCount: 0,
    ...sr,
  };
  return r;
};

const asMarkdownResult = (tsr: TestScenarioResult): TestScenarioMarkdownResult => {
  const fatalCount = tsr.stepResult.filter(
    (sr) => sr.runtimeError && sr.runtimeError.length > 0
  ).length;
  const errorCount = tsr.stepResult.filter(
    (sr) => sr.responseDiffResult && sr.responseDiffResult.length > 0
  ).length;
  let resultState: ResultState = "succeeded";
  if (fatalCount > 0) {
    resultState = "fatal";
  } else if (errorCount > 0) {
    resultState = "failed";
  } else {
    resultState = "succeeded";
  }

  const r: TestScenarioMarkdownResult = {
    testScenarioName: tsr.testScenarioName!,
    result: resultState,
    swaggerFilePaths: tsr.swaggerFilePaths,
    startTime: new Date(tsr.startTime!),
    endTime: new Date(tsr.endTime!),
    runId: tsr.runId!,
    fatalStepsCount: fatalCount,
    failedStepsCount: errorCount,
    warningStepsCount: 0,
    steps: tsr.stepResult.map(asMarkdownStepResult),
  };

  return r;
};

export const generateMarkdownReportHeader = (): string => "<h3>Azure API Test Report</h3>";
export const generateMarkdownReport = (testScenarioResult: TestScenarioResult): string => {
  const result = asMarkdownResult(testScenarioResult);
  const body = generateMarkdownReportView(result);
  return body;
};
export const generateJUnitCaseReport = (sr: StepResult): string => {
  const result = asMarkdownStepResult(sr);
  const body = generateJUnitCaseReportView(result);
  return body;
};
