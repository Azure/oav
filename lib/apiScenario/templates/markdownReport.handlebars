
{{#with this as |ts|}}
  <details>
    <summary>
      <code>
        {{renderSymbol ts.result}}
      </code>
      <b>
        {{renderScenarioTitle ts}}
      </b>
    </summary>
    <br />
    <ul>
      <li>
        Duration:
        {{renderDuration ts.startTime ts.endTime}}
      </li>
      <li>
        Swagger Files
        <ul>
          {{#each ts.swaggerFilePaths}}
            <li>
              <a target="_blank" href="{{renderPlain this}}">
                {{renderPlain this}}
              </a>
            </li>
          {{/each}}
        </ul>
      </li>
      <li>
        Run Id: {{renderPlain ts.runId}}
      </li>
    </ul>
    <b>
      &nbsp;&nbsp;&nbsp;&nbsp;API Test Steps:
    </b>
    <ol>
{{#each ts.steps}}
  {{#with this as |s|}}
<li>
  <code>
        {{renderSymbol s.result}}
  </code>
<b>
        {{renderStepTitle s}}
</b>
</li>
<ul>
    <li>
      StatusCode:
      {{renderPlain s.statusCode}}
    </li>
    <li>
      Duration:
      {{renderResponseTime s.responseTime}}
    </li>
    {{#if (shouldReportPayload s.payloadPath)}}
    <li>
      Payload:
      <a target="_blank" href="{{renderPlain s.payloadPath}}">
        {{renderPlain s.payloadPath}}
      </a>
    </li>
    {{/if}}
    <li>
      Correlation Id:
      {{renderPlain s.correlationId}}
    </li>
  </ul>

{{#if (shouldReportError s)}}
|{{renderWhitespace 27}}Rule{{renderWhitespace 27}}| Message |
|------|---------|
{{/if}}
            {{#each s.runtimeError}}
              {{#with this as |re|}}
| <code> {{renderSymbol "fatal"}} </code> {{renderFatalErrorCode re}} | {{renderFatalErrorDetail re}} |
              {{/with}}
            {{/each}}
            {{#each s.liveValidationResult.requestValidationResult.errors}}
            {{#with this as |rd|}}
| <code> {{renderSymbol "failed"}} </code> {{renderLiveValidationErrorCode rd}} | {{renderLiveValidationErrorDetail rd}} |
              {{/with}}
            {{/each}}
            {{#each s.liveValidationResult.responseValidationResult.errors}}
            {{#with this as |rd|}}
| <code> {{renderSymbol "failed"}} </code> {{renderLiveValidationErrorCode rd}} | {{renderLiveValidationErrorDetail rd}} |
              {{/with}}
            {{/each}}
          {{/with}}
<br />
{{/each}}

</ol>
</details>
{{/with}}