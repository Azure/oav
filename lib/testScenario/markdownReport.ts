import { readFileSync } from "fs";
import path from "path";
import Handlebars from "handlebars";
import moment from "moment";
import * as _ from "lodash";
import * as hd from "humanize-duration";
import { ResponseDiffItem, RuntimeError, StepResult, TestScenarioResult } from "./reportGenerator";

Handlebars.logger.log = function (level) {
  if (level >= Handlebars.logger.level) {
    console.log.apply(console, ([] as any).concat(["Handlebars: "], _.toArray(arguments)));
  }
};
// DEBUG: 0, INFO: 1, WARN: 2, ERROR: 3,
Handlebars.registerHelper("log", Handlebars.logger.log);
// Std level is 3, when set to 0, handlebars will log all compilation results
Handlebars.logger.level = 0;

const commonHelper = (opts: HelperOpts) => ({
  renderPlain: (s: string) => {
    return s;
  },
  renderUri: (s: string) => `${path.join(opts.swaggerRootDir, s)}`,
  renderSymbol: (result: ResultState) => {
    console.log("result=", result);
    return `${resultStateSymbol[result]}`;
  },
  renderScenarioTitle: (ts: TestScenarioMarkdownResult) =>
    `${ts.testScenarioName}: ${ts.fatalStepsCount} Fatals, ${ts.failedStepsCount} Errors`,
  renderStepTitle: (ts: TestScenarioMarkdownStepResult) =>
    `${ts.stepName}: ${ts.fatalStepsCount} Fatals, ${ts.failedStepsCount} Errors`,
  renderDuration: (start: Date, end: Date) =>
    `${hd.default(moment.duration(moment(end).diff(moment(start))).asMilliseconds())}`,
  shouldReportError: (sr: TestScenarioMarkdownStepResult) =>
    sr.failedStepsCount + sr.fatalStepsCount > 0,
  renderFatalErrorCode: (e: RuntimeError) => `${e.severity} ${e.code}`,
  renderFatalErrorDetail: (e: RuntimeError) => `${e.message}`, // todo add detail
  renderDiffErrorCode: (e: ResponseDiffItem) => `${e.severity} ${e.code}`,
  renderDiffErrorDetail: (e: ResponseDiffItem) => `${e.jsonPath} ${e.message}`, // todo add detail
});

type ResultState = keyof typeof ResultStateStrings;

const ResultStateStrings = {
  fatal: `Fatal`,
  failed: `Failed`,
  succeeded: `Succeeded`,
  warning: `Warning`,
};

export const resultStateSymbol: { [key in ResultState]: string } = {
  fatal: "❌",
  failed: "❌",
  succeeded: "️✔️",
  warning: "⚠️",
};

interface TestScenarioMarkdownStepResult {
  stepName: string;
  result: ResultState;
  exampleFilePath?: string;
  correlationId?: string;
  operationId: string;
  fatalStepsCount: number;
  failedStepsCount: number;
  warningStepsCount: number;
  runtimeError?: RuntimeError[];
  responseDiffResult?: ResponseDiffItem[];
}

interface TestScenarioMarkdownResult {
  testScenarioName: string;
  result: ResultState;
  swaggerFilePaths: string[];
  startTime: Date;
  endTime: Date;
  runId: string;
  fatalStepsCount: number;
  failedStepsCount: number;
  warningStepsCount: number;
  steps: TestScenarioMarkdownStepResult[];
}

interface MarkdownResult {
  title: string;
  scenarios: TestScenarioMarkdownResult[];
}

interface HelperOpts {
  swaggerRootDir: "root";
}

export const compileHandlebarsTemplate = <T>(fileName: string, opts: HelperOpts) => {
  const generationViewTemplate = readFileSync(
    path.join(__dirname, "templates", fileName)
  ).toString();
  const templateDelegate = Handlebars.compile<T>(generationViewTemplate, { noEscape: true });
  const helpers = commonHelper(opts);
  return (data: T) => templateDelegate(data, { helpers });
};

const generateView = compileHandlebarsTemplate<MarkdownResult>("markdownReport.handlebars", {
  swaggerRootDir: "root",
});

const stepIsFatal = (sr: StepResult) => sr.runtimeError && sr.runtimeError.length > 0;
const stepIsFailed = (sr: StepResult) => sr.responseDiffResult && sr.responseDiffResult.length > 0;

const asMarkdownStepResult = (sr: StepResult): TestScenarioMarkdownStepResult => {
  let result: ResultState = "succeeded";
  if (stepIsFatal(sr)) {
    result = "fatal";
  } else if (stepIsFailed(sr)) {
    result = "failed";
  }

  const r: TestScenarioMarkdownStepResult = {
    result,
    fatalStepsCount: sr.runtimeError ? sr.runtimeError.length : 0,
    failedStepsCount: sr.responseDiffResult ? sr.responseDiffResult.length : 0,
    warningStepsCount: 0,
    ...sr,
  };
  return r;
};

const asMarkdownResult = (tsr: TestScenarioResult): TestScenarioMarkdownResult => {
  const fatalCount = tsr.stepResult.filter(
    (sr) => sr.runtimeError && sr.runtimeError.length > 0
  ).length;
  const errorCount = tsr.stepResult.filter(
    (sr) => sr.responseDiffResult && sr.responseDiffResult.length > 0
  ).length;
  let resultState: ResultState = "succeeded";
  if (fatalCount > 0) {
    resultState = "fatal";
  } else if (errorCount > 0) {
    resultState = "failed";
  } else {
    resultState = "succeeded";
  }

  const r: TestScenarioMarkdownResult = {
    testScenarioName: tsr.testScenarioName!,
    result: resultState,
    swaggerFilePaths: tsr.swaggerFilePaths,
    startTime: new Date(tsr.startTime!),
    endTime: new Date(tsr.endTime!),
    runId: tsr.runId!,
    fatalStepsCount: fatalCount,
    failedStepsCount: errorCount,
    warningStepsCount: 0,
    steps: tsr.stepResult.map(asMarkdownStepResult),
  };

  return r;
};

export const generateMarkdownReport = (testScenarioResult: TestScenarioResult[]): string => {
  try {
    const result = testScenarioResult.map(asMarkdownResult);
    const body = generateView({
      title: "Azure API Test Report",
      scenarios: result,
    });
    console.log(JSON.stringify(result));
    return body;
  } catch (e) {
    console.error(e);
  }
  return "";
};
