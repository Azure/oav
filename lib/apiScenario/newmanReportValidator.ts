import * as path from "path";
import { findReadMe } from "@azure/openapi-markdown";
import { inject, injectable } from "inversify";
import { TYPES } from "../inversifyUtils";
import {
  LiveValidator,
  RequestResponseLiveValidationResult,
  RequestResponsePair,
} from "../liveValidation/liveValidator";
import { LiveRequest, LiveResponse } from "../liveValidation/operationValidator";
import { FileLoader } from "../swagger/fileLoader";
import { setDefaultOpts } from "../swagger/loader";
import { SwaggerExample } from "../swagger/swaggerTypes";
import { SeverityString } from "../util/severity";
import { getApiVersionFromFilePath, getProviderFromFilePath } from "../util/utils";
import { ApiScenarioLoaderOption } from "./apiScenarioLoader";
import { NewmanExecution, NewmanReport, Scenario, Step, StepRestCall } from "./apiScenarioTypes";
import { DataMasker } from "./dataMasker";
import { getJsonPatchDiff } from "./diffUtils";
import { JUnitReporter } from "./junitReport";
import { generateMarkdownReport } from "./markdownReport";
import { SwaggerAnalyzer } from "./swaggerAnalyzer";
import { VariableEnv } from "./variableEnv";

interface GeneratedExample {
  step: string;
  operationId: string;
  example: SwaggerExample;
}

export interface ApiScenarioTestResult {
  apiScenarioFilePath: string;
  readmeFilePath?: string;
  swaggerFilePaths: string[];
  tag?: string;

  // New added fields
  environment?: string;
  armEnv?: string;
  subscriptionId?: string;
  rootPath?: string;
  providerNamespace?: string;
  apiVersion?: string;
  startTime?: string;
  endTime?: string;
  runId?: string;
  repository?: string;
  branch?: string;
  commitHash?: string;
  baseUrl?: string;
  apiScenarioName?: string;
  stepResult: StepResult[];
}

export interface StepResult {
  statusCode: number;
  specFilePath?: string;
  exampleFilePath?: string;
  example?: SwaggerExample;
  operationId: string;
  runtimeError?: RuntimeError[];
  responseDiffResult?: ResponseDiffItem[];
  liveValidationResult?: RequestResponseLiveValidationResult;
  stepValidationResult?: any;
  correlationId?: string;
  stepName: string;
}

export interface RuntimeError {
  code: string;
  message: string;
  detail: string;
  severity: SeverityString;
}

export interface ResponseDiffItem {
  code: string;
  jsonPath: string;
  severity: SeverityString;
  message: string;
  detail: string;
}

export type ValidationLevel = "validate-request" | "validate-request-response";

export interface NewmanReportValidatorOption extends ApiScenarioLoaderOption {
  apiScenarioFilePath: string;
  reportOutputFilePath: string;
  markdown?: boolean;
  junit?: boolean;
  html?: boolean;
  specPathPrefix?: string;
  baseUrl?: string;
  runId?: string;
  validationLevel?: ValidationLevel;
  savePayload?: boolean;
  generateExample?: boolean;
  verbose?: boolean;
}

@injectable()
export class NewmanReportValidator {
  private scenario: Scenario;
  private testResult: ApiScenarioTestResult;
  private fileRoot: string;
  private liveValidator: LiveValidator;
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(
    @inject(TYPES.opts) private opts: NewmanReportValidatorOption,
    private fileLoader: FileLoader,
    private dataMasker: DataMasker,
    private swaggerAnalyzer: SwaggerAnalyzer,
    private junitReporter: JUnitReporter
  ) {
    setDefaultOpts(this.opts, {
      validationLevel: "validate-request-response",
      savePayload: false,
      generateExample: false,
      verbose: false,
    } as NewmanReportValidatorOption);
  }

  public async initialize(scenario: Scenario) {
    this.scenario = scenario;

    this.fileRoot =
      (await findReadMe(this.opts.apiScenarioFilePath)) ||
      path.dirname(this.opts.apiScenarioFilePath);

    this.testResult = {
      apiScenarioFilePath: path.relative(this.fileRoot, this.opts.apiScenarioFilePath),
      swaggerFilePaths: this.opts.swaggerFilePaths!,
      providerNamespace: getProviderFromFilePath(this.opts.apiScenarioFilePath),
      apiVersion: getApiVersionFromFilePath(this.opts.apiScenarioFilePath),
      runId: this.opts.runId,
      rootPath: this.fileRoot,
      repository: process.env.SPEC_REPOSITORY,
      branch: process.env.SPEC_BRANCH,
      commitHash: process.env.COMMIT_HASH,
      environment: process.env.ENVIRONMENT || "test",
      apiScenarioName: this.scenario.scenario,
      baseUrl: this.opts.baseUrl,
      stepResult: [],
    };

    await this.swaggerAnalyzer.initialize();

    this.liveValidator = new LiveValidator({
      fileRoot: "/",
      swaggerPaths: [...this.opts.swaggerFilePaths!],
    });
    await this.liveValidator.initialize();
  }

  public async generateReport(rawReport: NewmanReport) {
    await this.generateApiScenarioTestResult(rawReport);

    await this.outputReport();
  }

  private async generateApiScenarioTestResult(newmanReport: NewmanReport) {
    const variables = newmanReport.variables;
    this.testResult.startTime = new Date(newmanReport.timings.started).toISOString();
    this.testResult.endTime = new Date(newmanReport.timings.completed).toISOString();
    this.testResult.subscriptionId = variables.subscriptionId.value as string;
    for (const it of newmanReport.executions) {
      if (it.annotation === undefined) {
        continue;
      }
      if (it.annotation.type === "simple" || it.annotation.type === "LRO") {
        const runtimeError: RuntimeError[] = [];
        const matchedStep = this.getMatchedStep(it.annotation.step) as StepRestCall;

        // Runtime errors
        if (it.response.statusCode >= 400) {
          const error = this.getRuntimeError(it);
          runtimeError.push(error);
        }
        if (matchedStep === undefined) {
          continue;
        }

        const payload = this.convertToLiveValidationPayload(it);

        let responseDiffResult: ResponseDiffItem[] | undefined = undefined;
        const statusCode = `${it.response.statusCode}`;
        const exampleFilePath = `./examples/${matchedStep.operationId}_${statusCode}.json`;
        if (this.opts.generateExample || it.annotation.exampleName) {
          const generatedExample: SwaggerExample = {
            operationId: matchedStep.operationId,
            title: matchedStep.step,
            description: matchedStep.description,
            parameters: matchedStep._resolvedParameters!,
            responses: {
              [statusCode]: {
                headers: payload.liveResponse.headers,
                body: payload.liveResponse.body,
              },
            },
          };

          // Example validation
          if (this.opts.generateExample) {
            await this.fileLoader.writeFile(
              path.resolve(path.dirname(this.opts.reportOutputFilePath), exampleFilePath),
              JSON.stringify(generatedExample, null, 2)
            );
          }
          if (it.annotation.exampleName) {
            // validate real payload.
            responseDiffResult =
              this.opts.validationLevel === "validate-request-response"
                ? await this.exampleResponseDiff(
                    {
                      step: matchedStep.step,
                      operationId: matchedStep.operationId,
                      example: generatedExample,
                    },
                    matchedStep,
                    newmanReport
                  )
                : [];
          }
        }

        // Schema validation
        const correlationId = it.response.headers["x-ms-correlation-request-id"];
        if (this.opts.savePayload) {
          const payloadFilePath = `./payloads/${matchedStep.step}_${correlationId}.json`;
          await this.fileLoader.writeFile(
            path.resolve(path.dirname(this.opts.reportOutputFilePath), payloadFilePath),
            JSON.stringify(payload, null, 2)
          );
        }
        const liveValidationResult = await this.liveValidator.validateLiveRequestResponse(payload);

        this.testResult.stepResult.push({
          specFilePath: matchedStep.operation._path._spec._filePath,
          exampleFilePath: exampleFilePath,
          operationId: it.annotation.operationId,
          runtimeError,
          responseDiffResult: responseDiffResult,
          correlationId: correlationId,
          statusCode: it.response.statusCode,
          stepName: it.annotation.step,
          liveValidationResult: liveValidationResult,
        });
      }
    }
  }

  private convertToLiveValidationPayload(execution: NewmanExecution): RequestResponsePair {
    const request = execution.request;
    const response = execution.response;
    const liveRequest: LiveRequest = {
      url: request.url.toString(),
      method: request.method.toLowerCase(),
      headers: request.headers,
      body: request.body ? JSON.parse(request.body) : undefined,
    };
    const liveResponse: LiveResponse = {
      statusCode: response.statusCode.toString(),
      headers: response.headers,
      body: response.body ? JSON.parse(response.body) : undefined,
    };
    return {
      liveRequest,
      liveResponse,
    };
  }

  private convertPostmanFormat<T>(obj: T, convertString: (s: string) => string): T {
    if (typeof obj === "string") {
      return convertString(obj) as unknown as T;
    }
    if (typeof obj !== "object") {
      return obj;
    }
    if (obj === null || obj === undefined) {
      return obj;
    }
    if (Array.isArray(obj)) {
      return (obj as any[]).map((v) => this.convertPostmanFormat(v, convertString)) as unknown as T;
    }

    const result: any = {};
    for (const key of Object.keys(obj)) {
      result[key] = this.convertPostmanFormat((obj as any)[key], convertString);
    }
    return result;
  }

  private async exampleResponseDiff(
    example: GeneratedExample,
    matchedStep: Step,
    rawReport: NewmanReport
  ): Promise<ResponseDiffItem[]> {
    let res: ResponseDiffItem[] = [];
    if (matchedStep?.type === "restCall") {
      if (example.example.responses["200"] !== undefined) {
        Object.keys(matchedStep.variables).forEach((key) => {
          const paramName = `${matchedStep.step}_${key}`;
          matchedStep.responses = this.convertPostmanFormat(matchedStep.responses, (s) =>
            s.replace(`$(${key})`, `$(${paramName})`)
          );
        });
        res = res.concat(
          await this.responseDiff(
            example.example.responses["200"]?.body || {},
            matchedStep.responses["200"]?.body || {},
            rawReport.variables,
            `/200/body`,
            matchedStep.operation.responses["200"].schema
          )
        );
      }
    }
    return res;
  }

  private async responseDiff(
    resp: any,
    expectedResp: any,
    variables: any,
    jsonPathPrefix: string,
    schema: any
  ): Promise<ResponseDiffItem[]> {
    const env = new VariableEnv(variables);
    try {
      const ignoredKeys = [/\/body\/id/, /\/body\/location/, /date/i];
      const ignoreValuePattern = [/\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d(?:\.\d+)?Z?/];
      const expected = env.resolveObjectValues(expectedResp);
      const delta: ResponseDiffItem[] = getJsonPatchDiff(expected, resp, {
        includeOldValue: true,
        minimizeDiff: false,
      })
        .map((it: any) => {
          const ret: ResponseDiffItem = {
            code: "",
            jsonPath: "",
            severity: "Error",
            message: "",
            detail: "",
          };
          const jsonPath: string = it.remove || it.add || it.replace;
          const propertySchema = this.swaggerAnalyzer.findSchemaByJsonPointer(
            jsonPath,
            schema,
            resp
          );
          if (propertySchema["x-ms-secret"] === true) {
            return undefined;
          }
          if (it.remove !== undefined) {
            ret.code = "RESPONSE_MISSING_VALUE";
            ret.jsonPath = jsonPathPrefix + it.remove;
            ret.severity = "Error";
            ret.message = `The response value is missing. Path: ${
              ret.jsonPath
            }. Expected: ${this.dataMasker.jsonStringify(it.oldValue)}. Actual: undefined`;
          } else if (it.add !== undefined) {
            ret.code = "RESPONSE_ADDITIONAL_VALUE";
            ret.jsonPath = jsonPathPrefix + it.add;
            ret.severity = "Error";
            ret.message = `Return additional response value. Path: ${
              ret.jsonPath
            }. Expected: undefined. Actual: ${this.dataMasker.jsonStringify(it.value)}`;
          } else if (it.replace !== undefined) {
            ret.code = "RESPONSE_INCONSISTENT_VALUE";
            ret.jsonPath = jsonPathPrefix + it.replace;

            const paths = it.replace.split("/");
            while (paths.length > 1) {
              const path = paths.join("/");
              const sch = this.swaggerAnalyzer.findSchemaByJsonPointer(path, schema, resp);
              // Ignore diff when propertySchema readonly is true. When property is readonly, it's probably a random generated value which updated dynamically per request.
              if (sch.readOnly === true) {
                return undefined;
              }
              paths.pop();
            }
            ret.severity = "Error";
            ret.message = `The actual response value is different from example. Path: ${
              ret.jsonPath
            }. Expected: ${this.dataMasker.jsonStringify(
              it.oldValue
            )}. Actual: ${this.dataMasker.jsonStringify(it.value)}`;
          }
          ret.detail = this.dataMasker.jsonStringify(it);
          if (
            ignoredKeys.some((key) => ret.jsonPath.search(key) !== -1) ||
            (ret.code === "RESPONSE_INCONSISTENT_VALUE" &&
              typeof it.value === "string" &&
              ignoreValuePattern.some((valuePattern) => it.value.search(valuePattern) !== -1))
          ) {
            return undefined;
          }
          return ret;
        })
        .filter((it) => it !== undefined) as ResponseDiffItem[];
      return delta;
      // eslint-disable-next-line no-empty
    } catch (err) {}
    return [];
  }

  private getMatchedStep(stepName: string): Step | undefined {
    for (const it of this.scenario.steps ?? []) {
      if (stepName === it.step) {
        return it;
      }
    }
    return undefined;
  }

  private getRuntimeError(it: NewmanExecution): RuntimeError {
    const responseObj = this.dataMasker.jsonParse(it.response.body);
    return {
      code: "RUNTIME_ERROR",
      message: `statusCode: ${it.response.statusCode}, errorCode: ${responseObj?.error?.code}, errorMessage: ${responseObj?.error?.message}`,
      severity: "Error",
      detail: this.dataMasker.jsonStringify(it.response.body),
    };
  }

  private async outputReport(): Promise<void> {
    if (this.opts.reportOutputFilePath !== undefined) {
      console.log(`Write generated report file: ${this.opts.reportOutputFilePath}`);
      await this.fileLoader.writeFile(
        this.opts.reportOutputFilePath,
        JSON.stringify(this.testResult, null, 2)
      );
    }
    if (this.opts.markdown) {
      await this.fileLoader.appendFile(
        path.join(path.dirname(path.dirname(this.opts.reportOutputFilePath)), "report.md"),
        generateMarkdownReport(this.testResult)
      );
    }
    if (this.opts.junit) {
      await this.junitReporter.addSuiteToBuild(
        this.testResult,
        path.join(path.dirname(path.dirname(this.opts.reportOutputFilePath)), "junit.xml")
      );
    }
  }
}
