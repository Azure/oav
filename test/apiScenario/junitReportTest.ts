// Copyright (c) 2021 Microsoft Corporation
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
import { generateJUnitCaseReport } from "../../lib/apiScenario/markdownReport";
import { StepResult } from "../../lib/apiScenario/newmanReportValidator";

describe("junitTestReport", () => {
  it("Should generate junit case report", () => {
    const ts = {
      exampleFilePath:
        "Microsoft.Compute/preview/2020-09-30/test-scenarios/examples/CreateOrUpdateASimpleGalleryWithSharingProfile.json",
      operationId: "Galleries_CreateOrUpdate",
      runtimeError: [
        {
          code: "RUNTIME_ERROR_400",
          message:
            "code: BadRequest, message: Subscription 'db5eb68e-73e2-4fa8-b18a-46cd1be4cce5' is not registered with the feature Microsoft.Compute/SIGSharing. Please register the subscription before retrying the call.",
          severity: "Error",
          detail:
            '{\r\n  "error": {\r\n    "code": "BadRequest",\r\n    "message": "Subscription \'db5eb68e-73e2-4fa8-b18a-46cd1be4cce5\' is not registered with the feature Microsoft.Compute/SIGSharing. Please register the subscription before retrying the call."\r\n  }\r\n}',
        },
      ],
      responseDiffResult: [],
      stepValidationResult: [],
      correlationId: "cb309eda-ffb6-41f7-8a5e-42fc2a25aa5e",
      statusCode: 400,
      stepName: "Create or update a simple gallery with sharing profile.",
    } as StepResult;
    const body = generateJUnitCaseReport(ts);
    expect(body).toMatchSnapshot();
  });
});
