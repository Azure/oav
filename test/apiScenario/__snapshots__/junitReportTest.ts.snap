// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`junitTestReport Should generate junit case report 1`] = `
"###Create or update a simple gallery with sharing profile.: 1 Fatal Error(s)
| Rule | Message |
|------|---------|
|<code>‚ùå</code>[RUNTIME_ERROR_400]()|code: BadRequest, message: Subscription 'db5eb68e-73e2-4fa8-b18a-46cd1be4cce5' is not registered with the feature Microsoft.Compute/SIGSharing. Please register the subscription before retrying the call.|
"
`;
